File: frontend/src/services/departmentService.js
import api from "./api";

export const departmentService = {
  getAll: () => api.get("/departments"),
  getById: (id) => api.get(`/departments/${id}`),
  create: (payload) => api.post("/departments", payload),
  update: (id, payload) => api.put(`/departments/${id}`, payload),
  remove: (id) => api.delete(`/departments/${id}`),
};

export const positionService = {
  getAll: () => api.get("/positions"),
  getByDepartment: (departmentId) => api.get(`/positions/department/${departmentId}`),
  create: (payload) => api.post("/positions", payload),
  update: (id, payload) => api.put(`/positions/${id}`, payload),
  remove: (id) => api.delete(`/positions/${id}`),
};

File: backend/src/main/java/com/__Y2_S1_MLB_B10G1_05/demo/department/DepartmentService.java
package com.__Y2_S1_MLB_B10G1_05.demo.department;

import com.__Y2_S1_MLB_B10G1_05.demo.common.exception.BadRequestException;
import com.__Y2_S1_MLB_B10G1_05.demo.common.exception.ResourceNotFoundException;
import com.__Y2_S1_MLB_B10G1_05.demo.department.dto.DepartmentRequest;
import com.__Y2_S1_MLB_B10G1_05.demo.department.dto.DepartmentResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional
public class DepartmentService {

    private final DepartmentRepository departmentRepository;

    public List<DepartmentResponse> getAll() {
        return departmentRepository.findAll().stream()
                .map(DepartmentResponse::fromEntity)
                .toList();
    }

    public DepartmentResponse getById(Long id) {
        return DepartmentResponse.fromEntity(findEntity(id));
    }

    public Department findEntity(Long id) {
        return departmentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Department not found with id: " + id));
    }

    public DepartmentResponse create(DepartmentRequest request) {
        if (departmentRepository.existsByDepartmentName(request.getDepartmentName())) {
            throw new BadRequestException("Department already exists: " + request.getDepartmentName());
        }
        Department department = Department.builder()
                .departmentName(request.getDepartmentName())
                .description(request.getDescription())
                .build();
        return DepartmentResponse.fromEntity(departmentRepository.save(department));
    }

    public DepartmentResponse update(Long id, DepartmentRequest request) {
        Department department = findEntity(id);
        department.setDepartmentName(request.getDepartmentName());
        department.setDescription(request.getDescription());
        return DepartmentResponse.fromEntity(departmentRepository.save(department));
    }

    public void delete(Long id) {
        Department department = findEntity(id);
        departmentRepository.delete(department);
    }
}
